README.md
# wrapg
Wrapper around [_psycopg 3.x_](https://www.psycopg.org/psycopg3/docs/index.html) meant to make easy use of postgres using simple python functions. Primary focus is processing python data structures into and out of postgres.

Project is in infancy, _work in progress_.


## Table of Contents
* [Features](#features)
* [Installing Wrapg](#setup)
* [Usage](#usage)
* [Todo](#todo)
* [Acknowledgements](#acknowledgements)
* [Contact](#contact)
* [License](#license)


## Features
- Simple python functions to run postgres sql via python. see Usage section for more details on list of functionality.
    - upsert & insert_ignore functions included
        - 'use_index=True' automatically creates index
        - 'use_index=False' to upsert without using index (slower)
    - copy_from_csv() function to follow postgres COPY protocol *(today only csv is avail)*
- Pass/Retrieve various python data structutes via underlining sql functions
    - supports pandas dataframe out of box
- Auto import default postgres connection parameters via .env
    - saves on repeating code to specify connection
    - overide default connection parameters with kwargs in each function if needed
- Use of sql functions with certain parametes. *(work in progress, today mosty use Date() to type cast in keys parameters)*


## Installing Wrapg
Wrapg is available on PyPI:
```
pip install wrapg
```

Dependencies:
- python 3.10+
- [psycopg[binary]>=3.0.11+](https://www.psycopg.org/psycopg3/docs/index.html)
- [pandas>=1.4.2+](https://pandas.pydata.org/docs/index.html)


## Usage 

### Connection
Before you get started is is recommended to create .env file with below connection parameters.
Wrapg will auto-import default connectionn parameters and make all functions ready to be executed.  
The .env file should contain below specific name variables
```
# Database config
PG_HOST = "localhost"
PG_USER = "postgres"
PG_PASSWORD = "mypass"
PG_PORT = 1234
PG_DBNAME = "supers"
```

> *Any connection variable can be overwritten in each function call if needed per [postgres connection parameter key words](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS).*

Below will overide default PG_DBNAME from 'supers' database to the 'sales' database for specific function call.
```
qry="SELECT * FROM customers" 
wrapg.query(raw_sql=qry, conn_kwargs={'dbname': 'sales'})

```

### Create Table
Function to help create table from python.

```
cols = dict(id="serial", name="varchar(75) unique not null", age="int")
wrapg.create_table(table="villian", columns=cols)
```

### Insert
Insert function using list of dictionaries.  
*('info' can be in pandas dataframe form.)*
```
info = [{'name': 'Peter Paker', superhero: 'Spider-man', 'email': 'webhead@gmail.com'},
{'name': 'Bruce Wayne', superhero: 'Batman', 'email': 'bwayne@gmail.com'}]

wrapg.insert(data=info, table="superhero")
```

### Update
Easily call sql update.  
- If rows with matching keys exist, update row values.
- The columns/info that is not provided in the 'data' retain their original values.
- keys parameter must be specified in function.

```
new_email = {superhero: 'Spider-man', 'email': 'spidey@gmail.com'}
wrapg.update(data=new_email, table="superhero", keys=["superhero"])
```


### Upsert
Easily call sql upsert.  
- Add a row into specified table if the row with specified keys does not already exist. 
    - If rows with matching keys parameter exist, update row values.
- Automatically creates unique index if one does not exist for keys provided when use_index=True (Default)
    - If use_index=False, auto creation of index will not occur and operation will first try to update record, then insert (slower)

```
record = {'name': 'Steve Rogers', superhero: 'Captian America', 'email': 'cap@gmail.com'}
wrapg.upsert(data=record, table="superhero", keys=["email"], use_index=True)
```

### Insert Ignore
Easily call sql insert ignore. 
- Add a row into specified table if the row with specified keys does not already exist.
    - If rows with matching keys exist no change is made.
- Automatically creates unique index if one does not exist for keys provided.
    - *(Future give option to turn off auto index)*

```
record = {'name': 'Dr Victor von Doom', villian: 'Dr Doom', 'email': 'doom@gmail.com'}
wrapg.insert_ignore(data=record, table="villian", keys=["email"], use_index=True)
```

### Query
For more complicated sql not covered by a specific function, one can use query() function to pass raw sql. 

```
qry="SELECT COUNT(DISTINCT alarm), locid, "Date" FROM metrics
WHERE "Date"='2020-08-02'
GROUP BY locid
ORDER BY COUNT(alarm) DESC" 

wrapg.query(raw_sql=qry)
```

## Todo
- Handle other operators other than '='; >, <, <>, in, between, like?
- insert_ignore() without index
- Implement create_index(), distinct(), drop_index()
- Table manupulation drop_column(), add_column(), delete_table()
- Handle JSON, ITERATOR?
- Add more tests
- Optimize code after it is all working


## Acknowledgements
This project built on great work by [psycopg 3](https://www.psycopg.org/psycopg3/docs/index.html) and was inspired by [_dataset_](https://dataset.readthedocs.io/en/latest/) 


## Contact
Wrapped by [_jturnercode_](https://github.com/jturnercode)


## License
- MIT




HISTORY.md
# Release History

0.2.4 (2022-08-19)
-------------------
**Bugfixes**
- Fix snippet module refernce to iterable_difference. from wrapg import util vs from util import iterable_difference


0.2.3 (2022-08-14)
-------------------
**Improvements**
- Add more tests for wrapg functions
- Add exclude_update parameter to update() and upsert() functions

**Bugfixes**
- Fix colname_snip() to properly cast types if date() passed; it was affecting upsert with index=False
- Fixed exisitng snippet tests due to type casting issue on colname_snip()


0.2.2 (2022-05-07)
-------------------
**Improvements**
- Add insert_snip() to resuse insert code
- Improve regex to better detect functions in column
- Add 'use_index' to upsert(); set to False to upsert without use of unique index
- Add more tests on util.py and snippet.py

**Bugfixes**
- Fix sqlfunc() to properly cast a timestamp to date()
- Fix update() with ' AND '


0.2.1 (2022-04-21)
-------------------
**Improvements**
- Add code for delete() which takes dictionary for where clause parmeters
    - The column name can be wrapped with sql function like upsert and update functions
- Add clear_table() to quickly clear all data in specified table


0.2.0 (2022-04-13)
-------------------
**Improvements**
- Add code to accept sql functions on column keys for upsert() and ignore_insert() functions.
ie. upsert(keys=['name', 'Date(ts_column)'])
- Change .env PG_USERNAME variable to PG_USER to better match postgres user connection parameter.
- Added some examples to readme.md

**Bugfixes**
- Escape create_table() column names


0.1.4 (2022-04-11)
-------------------
**Improvements**


**Bugfixes**
- Upsert() not working due to bad code.

tests/common.py
from wrapg import wrapg
from typing import Iterable


def drop_index(table: str, keys: Iterable):
    """
    Drop index created by wrapg for specified table.
    index format: tablename_keys_uix

    Args:
        table (str): tablename
    """
    key_str = "_".join(keys)

    try:
        drop_indx = f'DROP INDEX "{table}_{key_str}_uix";'
        wrapg.query(raw_sql=drop_indx)
    except Exception as e:
        # print('error>>>>>: ', e)
        pass


def clear_table(table: str):
    """Clear all records in specified table.

    Args:
        table (str): _description_
    """

    wrapg.clear_table(table=table)


tests/test_upsert.py
import os
from datetime import datetime
from wrapg import wrapg
import common


test_table = os.environ.get("TEST_TABLE")


def test_upsert_noindex():

    # ================================================
    #              Upsert() no index
    #
    # - uses sql date() type cast function in key
    # ================================================

    # clear index specified index and table records
    common.drop_index(test_table, keys=["Date(ts)"])
    common.clear_table(test_table)

    insert_data = [
        {
            "age": 3,
            "superhero": "Spider-man",
            "bike": "BMX",
            "name": "Matthew",
            "ts": datetime(2022, 4, 1, 7, 0),
        }
    ]

    # first insert data to check update in upsert works properly
    wrapg.insert(data=insert_data, table=test_table)

    upsert_data = [
        {
            "age": 7,
            "superhero": "Gold Spider",
            "bike": "BMX",
            "name": "Matthew",
            "ts": datetime(2022, 4, 1, 11, 0),
        },
        {
            "age": 100,
            "bike": "New Bike",
            "name": "James",
            "superhero": "Batman",
            "ts": datetime(2022, 8, 1, 8, 0),
        },
    ]

    # data to upsert using date(ts) key
    wrapg.upsert(data=upsert_data, table=test_table, keys=["Date(ts)"], use_index=False)

    # check updated matthew record
    qry = f"SELECT * FROM {test_table} WHERE name='Matthew'"
    result = wrapg.query(raw_sql=qry)
    record = list(result)[0]

    # check update in upsert()
    assert record == upsert_data[0]

    # check inserted james record
    qry = f"SELECT * FROM {test_table} WHERE name='James'"
    result = wrapg.query(raw_sql=qry)
    record = list(result)[0]

    # check insert in upsert()
    assert record == upsert_data[1]


def test_upsert_index():

    # ================================================
    #              Upsert() with index
    #
    # - uses sql type case function in key
    # ================================================

    # clear index specified index and table records
    common.drop_index(test_table, keys=["Date(ts)"])
    common.clear_table(test_table)

    insert_data = [
        {
            "age": 3,
            "superhero": "Spider-man",
            "bike": "BMX",
            "name": "Matthew",
            "ts": datetime(2022, 4, 1, 7, 0),
        }
    ]

    # insert test data
    wrapg.insert(data=insert_data, table=test_table)

    upsert_data = [
        {
            "age": 7,
            "superhero": "Gold Spider",
            "bike": "BMX",
            "name": "Matthew",
            "ts": datetime(2022, 4, 1, 11, 0),
        },
        {
            "age": 100,
            "bike": "New Bike",
            "name": "James",
            "superhero": "Batman",
            "ts": datetime(2022, 8, 1, 8, 0),
        },
    ]

    # data to upsert using date(ts) key
    wrapg.upsert(data=upsert_data, table=test_table, keys=["Date(ts)"], use_index=True)

    # check updated matthew record
    qry = f"SELECT * FROM {test_table} WHERE name='Matthew'"
    result = wrapg.query(raw_sql=qry)
    record = list(result)[0]

    # check update in upsert()
    assert record == upsert_data[0]

    # check inserted james record
    qry = f"SELECT * FROM {test_table} WHERE name='James'"
    result = wrapg.query(raw_sql=qry)
    record = list(result)[0]

    # check insert in upsert()
    assert record == upsert_data[1]

    # clean up for other tests
    common.drop_index(test_table, keys=["Date(ts)"])
    common.clear_table(test_table)


tests/test_util_funcs.py
from wrapg import util


# Note: run test using -m flag
# pipenv run python -m pytest -vv


def test_uniform_dict_keys():
    data1 = (
        {"num": 30, "data": "thirty", "bike": "brown", "name": "pete"},
        {"num": 60, "data": "sixty", "bike": "green", "name": "joe"},
        {"num": 80, "data": "eighty", "bike": "red"},
        {"num": 90, "data": "ninety", "bike": "candy red", "name": "sr"},
    )

    data2 = (
        {"num": 30, "data": "thirty", "bike": "brown", "name": "pete"},
        {"num": 60, "data": "sixty", "bike": "green", "name": "joe"},
        {"num": 80, "data": "eighty", "bike": "red", "name": "burt"},
        # note order should not matter
        {"num": 90, "bike": "candy red", "name": "sr", "data": "ninety"},
    )
    assert util.uniform_dict_keys(data1) != 1
    assert util.uniform_dict_keys(data2) == 1


def test_check_all_dicts():
    not_alldicts = (
        {"num": 30, "data": "thirty", "bike": "brown", "name": "Pete"},
        {"num": 80, "data": "eighty", "name": "Red"},
        (1, 2, 3),
    )

    all_dicts = (
        {"num": 30, "data": "thirty", "bike": "brown", "name": "Matthew"},
        {"num": 80, "bike": "Red", "name": "Ethan", "data": "eighty"},
    )

    # data_transform() returns (column names, )
    # test returned columns
    # TODO: How do i check BaseException?
    # assert util.check_all_dicts(not_alldicts) == 'BaseException: Iterable has mixed types, expected Iterable[dictionaries]'
    assert util.check_all_dicts(all_dicts) is True


def test_data_transform():
    tup_dict = (
        {"num": 30, "data": "thirty", "bike": "brown", "name": "Pete"},
        {"num": 80, "data": "eighty", "name": "Red"},
    )

    test_dict = {"num": 80, "data": "eighty", "name": "Red"}

    # data_transform() returns (column names, data, uniform)
    # test returned columns
    assert util.data_transform(tup_dict) == (tuple(tup_dict[0]), tup_dict, 2)
    assert util.data_transform(test_dict) == (tuple(test_dict), (test_dict,), 1)
    # TODO: Check more data structures


wrapg/util.py
from collections.abc import Iterable
import pandas as pd
from numpy import nan


def check_all_dicts(iterable_dict: Iterable[dict]):
    """Check if Iterable contains all dictionaries

    Args:
        iterable_dict (Iterable[dict]): Iterable of dictionaries
    """
    # Check if dict
    def check_dict(d):
        return isinstance(d, dict)

    # Check if all instances are type dict, return True or False
    all_dict = all(map(check_dict, iterable_dict))
    # print(all_dict)

    if not all_dict:
        raise BaseException("Iterable has mixed types, expected Iterable[dictionaries]")

    return True


def uniform_dict_keys(iterable_dict: Iterable[dict]) -> int:
    """
    Function evaluates if Iterable of dictionaries
    has same keys for each dictionary.
    If true, this will allow for one query
    to process data. ie one query to insert/update/etc
    all rows vs having to create a query
    for each dictionary (non uniform key/pair data)

    Args:
        iter_dict (_type_): Iterable of dictionaries
        representing data to be processed into database

    Return:
     Length of unique sets of keys. If returns 1,
     then this indicates all dicts in list
     have the same keys (uniform)
    """

    def key_sort(dict) -> tuple:
        """Return sorted dictionary keys for each dict"""
        # sorted(dict) returns keys
        # Need a tuple to process set()
        return tuple(sorted(dict))

    # Compare all tuples of keys and apply set() to identify unique key combinations
    # If len(set) = 1, then all keys same for all dictionaries->uniform data structure
    keys = set(map(key_sort, iterable_dict))

    return len(keys)


def data_transform(data_structure):
    """Internal function checks passed data structure and
    returns tuple of columns and Iterable of rows(dictionaries)

    Args:
        data_structure (Any): data needing to be inserted/
        updated/etc into postgres (type: dataframe,
        list/tuple of dict, dict)

    Returns:
        column, row, uniform: tuple(column_names), Iterable(dict), int
        uniform = 1 indicates all dictionaries have same keys
    """

    # =================== TODO ===================
    # TODO: handle json data, named tuple?
    # TODO: handle iterator?
    # TODO: test if returning rows of tuples faster to process?

    # structural pattern matching for data_structure passed
    match data_structure:
        case pd.DataFrame():
            """
            Dataframe is a uniform data structure, no varying
            columns for each row; missing values are converted
            to None for postgres
            """
            # print("type -> dataframe")

            # =============== df to tuple(not used) ===============
            # Need to replace all NaN to None, pg sees nan as str
            # df = data_structure.replace(nan, None)
            # # return named tuple
            # rows = tuple(df.itertuples(index=False, name=None))

            columns = tuple(data_structure.columns)
            # in case a df with nan is passed, None needed for sql
            df = data_structure.replace(nan, None)
            # returns list of dictionaries
            rows = df.to_dict(orient="records")
            uniform = 1

            # print(rows)
            return columns, rows, uniform

        case list() | tuple():
            # print("type -> list/tuple of dictionaries")

            # =========== iterable[dict] to df not used============
            # !ISSUE: Convert dict to df may update some columns
            # !for non-uniform case to None, unintentionally
            # !(non-uniform keys in all dicts to df)
            # df = pd.DataFrame(data_structure)
            # columns = df.columns
            # FIXME: below will convert intergers to float if column has nan, fix!
            # df = df.replace(nan, None)
            # rows = list(df.itertuples(index=False, name=None))

            # Check all instances in data structure are dictionaires
            if check_all_dicts(data_structure):

                # Return tuple(dict.keys of first instance in Iterable)
                # If not uniform program will iterate over instance
                # if uniform, checking first instance is good enough
                columns = tuple(data_structure[0])

                # Check if all dictionaries have same keys
                uniform = uniform_dict_keys(data_structure)

                return columns, data_structure, uniform

        case dict():
            # print("type -> dictionary")

            # return tuple (not used for now)
            # rows = [tuple(data_structure.values())]

            # return keys of dict in a tuple
            columns = tuple(data_structure)
            # tuple of one dictionary
            row = (data_structure,)
            uniform = 1

            return columns, row, uniform

        case []:
            raise ValueError(
                f"Empty list passed, must contain at least one dictionary."
            )

        case _:
            raise ValueError(f"Unsupported data structure passed.")



def iterable_difference(minuend: Iterable, subtrahend: Iterable) -> tuple:
    """Used to get the difference between two iterables or sequences.
    ie minuend - subtrahend = difference

    Args:
        minuend (Iterable): _description_
        subtrahend (Iterable): _description_
    """


    return tuple(m for m in minuend if m not in subtrahend)

wrapg/snippet.py
import re
from typing import Iterable
from psycopg import sql, connect
from wrapg import util


# TODO: today use date() to type cast; change regex to '::' for type cast vs '()' for true funcs
# Regex to seperate sql_func from column name
__compiled_pattern = re.compile(pattern=r"(\w*)\((\w*)\)")


# =================== Snippet Util Functions ===================


def check_for_func(sequence: Iterable) -> bool:
    """Used to determine if a list or tuple of
    columns passed into sql function has
    parethesis '()' which indicate a function
    that needs to be parsed out

    Args:
        sequence (Iterable): list/tupe of column names

    Returns:
        bool: True if function found
    """
    # make all elements strings
    seq_of_str = map(str, sequence)
    # combine seq of str into one long str
    combined_seq = "".join(seq_of_str)

    # return true or false
    return "(" in combined_seq


def colname_snip(sqlfunc_colname: tuple):
    """Return escaped sql snippet, accomodate column names
    wrapped by sql functions.

    Args:
        column_detail (str | tuple): column_name as str
        or tuple (sql_func, column_name)

    Returns:
        Composed: snippet of sql statment
    """
    sqlfunc, colname = sqlfunc_colname

    # return snippet of sql func wrapping column name
    if sqlfunc is None:
        # return escaped column name
        return sql.SQL("{}").format(
            sql.Identifier(colname),
        )

    # Returns sql func format; Keep for later use for true functions
    return sql.SQL("{}({})").format(
        sql.SQL(sqlfunc),
        sql.Identifier(colname),
    )

    # Returns sql type cast (::) syntax
    # return sql.SQL("{}::{}").format(
    #     sql.Identifier(colname),
    #     sql.SQL(sqlfunc),
    # )


# =================== Unique Index Snippet ===================


def create_unique_index(table, keys):

    # Note name will include parenthsis if passed
    # in unique index name
    uix_name = f'{table}_{"_".join(keys)}_uix'

    # If sql function in the any key
    if check_for_func(keys):

        # Return tuple of (sql func, colname)
        sqlfunc_keys = map(get_sqlfunc_colname, keys)

        # sql snippet to create unique index
        return sql.SQL("CREATE UNIQUE INDEX {} ON {} ({});").format(
            sql.Identifier(uix_name),
            sql.Identifier(table),
            sql.SQL(", ").join(map(colname_snip, sqlfunc_keys)),
        )

    # Sql snippet to create unique index
    return sql.SQL("CREATE UNIQUE INDEX {} ON {} ({});").format(
        sql.Identifier(uix_name),
        sql.Identifier(table),
        sql.SQL(", ").join(map(sql.Identifier, keys)),
    )


# =================== Upsert Snippets ===================

# Function to compose col=excluded.col sql for update
def exclude_sql(col):
    return sql.SQL("{}=EXCLUDED.{}").format(
        sql.Identifier(col),
        sql.Identifier(col),
    )


def upsert_snip(
    table: str, columns: Iterable, keys: Iterable, exclude_update: Iterable = None
):

    update_columns = columns

    # if exclude columns from update then determine update_columns
    if exclude_update:
        update_columns = util.iterable_difference(columns, exclude_update)

    # if sql function in the any key
    if check_for_func(keys):

        # Return tuple of (sql func, key)
        sqlfunc_keys = map(get_sqlfunc_colname, keys)

        # Sql snippet to upsert
        return sql.SQL(
            "INSERT INTO {} ({}) VALUES ({}) ON CONFLICT ({}) DO UPDATE SET {};"
        ).format(
            sql.Identifier(table),
            sql.SQL(", ").join(map(sql.Identifier, columns)),
            sql.SQL(", ").join(map(sql.Placeholder, columns)),
            # conflict target
            sql.SQL(", ").join(map(colname_snip, sqlfunc_keys)),
            # set new values
            sql.SQL(", ").join(map(exclude_sql, update_columns)),
        )

    # Sql snippet to upsert
    return sql.SQL(
        "INSERT INTO {} ({}) VALUES ({}) ON CONFLICT ({}) DO UPDATE SET {};"
    ).format(
        sql.Identifier(table),
        sql.SQL(", ").join(map(sql.Identifier, columns)),
        sql.SQL(", ").join(map(sql.Placeholder, columns)),
        # conflict target
        sql.SQL(", ").join(map(sql.Identifier, keys)),
        # set new values
        sql.SQL(", ").join(map(exclude_sql, update_columns)),
    )


# =================== Insert_ignore Snippet ===================


def insert_ignore_snip(table: str, columns, keys):

    # If sql function in the any key
    if check_for_func(keys):

        # Return tuple (sql func, key)
        sqlfunc_keys = map(get_sqlfunc_colname, keys)

        # Sql snippet to insert ignore
        return sql.SQL(
            "INSERT INTO {} ({}) VALUES ({}) ON CONFLICT ({}) DO NOTHING"
        ).format(
            sql.Identifier(table),
            sql.SQL(", ").join(map(sql.Identifier, columns)),
            sql.SQL(", ").join(map(sql.Placeholder, columns)),
            # conflict target
            sql.SQL(", ").join(map(colname_snip, sqlfunc_keys)),
        )

    # Sql snippet to insert ignore
    return sql.SQL(
        "INSERT INTO {} ({}) VALUES ({}) ON CONFLICT ({}) DO NOTHING"
    ).format(
        sql.Identifier(table),
        sql.SQL(", ").join(map(sql.Identifier, columns)),
        sql.SQL(", ").join(map(sql.Placeholder, columns)),
        # conflict target
        sql.SQL(", ").join(map(sql.Identifier, keys)),
    )


# =================== Delete Snippet ===================

# Note: Seperated from where_snip() to possibly reuse
# on other dictionaries needing to be composed for other
# snippets in future
def compose_key_value(key_value: tuple) -> tuple:
    """Take key_value tuple from dictionary via .items()
    and create a composed key_value tuple for use in creating
    snippets.

    Args:
        key_value (tuple): key value pair from dictionary

    Returns:
        tuple: composed key value
    """

    colname, value = key_value

    # TODO: ? add >, <, <>, etc to value to process data other than '='
    # composed value to literal type
    composed_value = sql.Literal(value)

    # Check if colname has a function
    if "(" in colname:
        # pattern = r"(\w*)\((\w*)\)"
        result = re.search(__compiled_pattern, colname)

        # Extract matching values of all groups
        # this is done to escape column name
        sqlfunc = result.group(1).upper()
        column = result.group(2)

        composed_column = sql.SQL("{}({})").format(
            sql.SQL(sqlfunc), sql.Identifier(column)
        )

        return composed_column, composed_value

    # composed column if no func found
    composed_column = sql.Identifier(colname)
    return composed_column, composed_value


# Note: Seperated from compose_key_value() as there
# may be need for compose other dictionaries in future
def where_snip(colname_value: tuple):
    """Represent where clause colname=value

    Args:
        colname_value (tuple): _description_

    Returns:
        _type_: _description_
    """

    colname, value = colname_value

    # return where clause snip for final snippet
    return sql.SQL("{}={}").format(colname, value)


def delete_snip(table: str, where: dict):
    """Base sql snippet for delete().

    Args:
        table (str): database table name
        where (dict): dict(colname=value) that
        filters value to remove from table

    Returns:
        _type_: _description_
    """

    # Pass key-value tuple to compose_key_value()
    composed_where = map(compose_key_value, where.items())

    return sql.SQL("DELETE FROM {} WHERE {};").format(
        sql.Identifier(table),
        sql.SQL(" AND ").join(map(where_snip, composed_where)),
    )


# =================== Update Snippets ===================


def get_sqlfunc_colname(colname: str):
    """
    Extract sql function from column name.
    Return (sql function, column name)
    If no sql function return None in tuple.
    Used on Iterable of strings via map().

    Args:
        colname (str): column name

    Returns:
        tuple: (None, colname) or (sqlfunc, colname)
    """
    # Check if colname has a function
    if "(" in colname:
        try:
            # pattern = r"(\w*)\((\w*)\)"
            result = re.search(__compiled_pattern, colname)

            # Extract matching values of all groups
            # this is done to escape column name
            sqlfunc = result.group(1).upper()
            colname = result.group(2)

            return sqlfunc, colname
        except:
            print("Invalid column name passed!")
    return None, colname


# function used to map passed dictionary values to column names
# TODO: Truly distinguish between sql func and type cast; limited use today
def colname_placeholder_snip(sqlfunc_colname: tuple):

    """
    Return sql snip for where and set clauses with
    named placeholder for value;
    ex. "SET col=value, col2=value2
    colname=%(colname)s or sqlfunc(colname)=%(colname)s

    Args:
        sqlfunc_colname (tuple): (sqlfunc, colname)

    Returns:
        Composable: colname=%(colname)s
    """
    # unpack sqlfunc_colname to determine colname placeholder syntax
    sqlfunc, colname = sqlfunc_colname
    
    if sqlfunc is None:
        return sql.SQL("{}={}").format(
            sql.Identifier(colname),
            sql.Placeholder(colname),
        )

    # NOTE: KEEP FOR NOW IN CASE WANT TO PROCESS TRUE FUNCTIONS VS TYPE CAST FUNCTIONS
    # return sql.SQL("{}({})={}").format(
    #     sql.SQL(sqlfunc),
    #     sql.Identifier(colname),
    #     sql.Placeholder(colname),
    # )

    # type cast if func found, Update future to handle better
    return sql.SQL("{}::{}={}::{}").format(
        sql.Identifier(colname),
        sql.SQL(sqlfunc),
        sql.Placeholder(colname),
        sql.SQL(sqlfunc),
    )


def update_snip(
    table: str, columns: Iterable, keys: Iterable, exclude_update: Iterable = None
):

    # if exclude columns from update then determine update_columns
    if exclude_update:
        columns = util.iterable_difference(columns, exclude_update)

    columns = map(get_sqlfunc_colname, columns)
    keys = map(get_sqlfunc_colname, keys)

    return sql.SQL("UPDATE {} SET {} WHERE {};").format(
        sql.Identifier(table),
        sql.SQL(", ").join(map(colname_placeholder_snip, columns)),
        sql.SQL(" AND ").join(map(colname_placeholder_snip, keys)),
    )


def insert_snip(table: str, columns: Iterable):

    return sql.SQL("INSERT INTO {} ({}) VALUES ({})").format(
        sql.Identifier(table),
        sql.SQL(", ").join(map(sql.Identifier, columns)),
        sql.SQL(", ").join(map(sql.Placeholder, columns)),
    )


if __name__ == "__main__":
    # dev testing, remove later
    import os
    from timeit import timeit

    # print("SNIPPET.PY")

    conn_import: dict = {
        "user": os.environ.get("PG_USER"),
        "password": os.environ.get("PG_PASSWORD"),
        "host": os.environ.get("PG_HOST"),
        "dbname": os.environ.get("PG_DBNAME"),
        "port": os.environ.get("PG_PORT"),
    }

    # Connect to an existing database
    with connect(**conn_import) as conn:

        dtest = {"name": "bos", "age": 34, "Date(ts)": "2022-04-19"}
        # dtest = {"name": "bos", "age": 34, "ts": "2022-04-19"}

        # tests
        # snipp = create_unique_index("mytable", ["name", "Date(ts)"])

        # timeit(delete_snip, "mytable", dtest)
        # def a():
        # snipp = delete_snip("mytable", dtest)
        # print(snipp.as_string(conn))

        # snipp = delete_snip("mytable", dtest)

        snipp = update_snip(
            table="mytable", columns=["a", "b", "c"], keys=["Date(b)", "d"]
        )
        print(snipp.as_string(conn))

        # Code used to time speed of functions
        # print(timeit(stmt='delete_snip("mytable", dtest)', number=1000, globals=globals())/1000)
        # print(timeit(stmt='delete_snip("mytable", dtest)', setup='from __main__ import delete_snip, dtest', number=1))
        # print(timeit(stmt='a()', setup='from __main__ import a', number=100))

        conn.close()
